packages/core/src/plugins/types.ts

```diff
@@ -1,26 +1,47 @@
-import { LogEntry } from '../queues';
+import { LogEntry, Logger, TContextBase } from '../logger';
+import { PluginManager } from './manager';
 
-export interface Plugins {
-  [id: string]: Plugin;
+export interface Plugins<
+  TContext extends TContextBase = { processId: string },
+  TGlobalContext extends Record<string, any> = {},
+> {
+  [id: string]: Plugin<Partial<TContext>, TGlobalContext>;
 }
 
 export type PluginId = Exclude<keyof Plugins, number | symbol>;
 
-export interface Plugin {
+export interface Plugin<
+  TContext extends TContextBase = { processId: string },
+  TGlobalContext extends Record<string, any> = {},
+> {
   readonly id: string;
+  readonly entryFormat: string;
+  injectLogger(
+    logger: Logger<Partial<TContext>, TGlobalContext>,
+    pluginManager: PluginManager,
+  ): void;
 }
 
-export interface FormatterPlugin extends Plugin {
-  getEntryLabel(entry: LogEntry): string;
-  getEntryTitle(entry: LogEntry): string;
-  formatHtmlEntry(entry: LogEntry): string;
-  formatConsoleEntry(entry: LogEntry): string;
+export interface FormatterPlugin<
+  TContext extends TContextBase = { processId: string },
+  TGlobalContext extends Record<string, any> = {},
+> extends Plugin<Partial<TContext>, TGlobalContext> {
+  readonly entryFormat: string;
+  readonly handlerSupport: string;
+  getEntryLabel(entry: LogEntry<Partial<TContext>, TGlobalContext>): string;
+  getEntryTitle(entry: LogEntry<Partial<TContext>, TGlobalContext>): string;
+  formatEntry(entry: LogEntry<Partial<TContext>, TGlobalContext>): any;
 }
 
-export function isFormatterPlugin(plugin: Plugin): plugin is FormatterPlugin {
+export function isFormatterPlugin<
+  TContext extends TContextBase = { processId: string },
+  TGlobalContext extends Record<string, any> = {},
+>(
+  plugin: Plugin<Partial<TContext>, TGlobalContext>,
+): plugin is FormatterPlugin<Partial<TContext>, TGlobalContext> {
   return (
-    typeof (plugin as any).getEntryLabel === 'function' &&
-    typeof (plugin as any).formatHtmlEntry === 'function' &&
-    typeof (plugin as any).formatConsoleEntry === 'function'
+    typeof (plugin as FormatterPlugin<Partial<TContext>, TGlobalContext>).getEntryLabel ===
+      'function' &&
+    typeof (plugin as FormatterPlugin<Partial<TContext>, TGlobalContext>).formatEntry === 'function'
   );
```

Comment:
tady je podle mě "divný" typecastovat to rovnou na `FormatterPlugin` - protože v tomhle místě ještě _nevíš_, jestli je to `FormatterPlugin` - a když to na něj typecastuješ, tak ta podmínka bude vždycky true (protože `typeof formatterPlugin.getEntryLabel` bude vždycky `function`), což jednak třeba IDE (i bez linterů etc) může označit za zbytečnou podmínku / zjednodušitelnej výraz, a některý lintery / nastavení linteru to taky můžou pohejtit

Reply:
<img width="460" alt="image" src="https://user-images.githubusercontent.com/104579053/169795291-4d84aa6b-2810-4a15-a66b-8e3cb1575a05.png">

Typecast ti udela veci jen to typescript, v runtime/po buildu realne neovlivni vubec nic

Reply:
To jo, ale IDE ten typescript vidí a bude se díky tomu typecastu ptát, zda nelze podmínku zjednodušit.. nebo možná ne, to nevím, občas to navrhuje a občas ne, ale každopádně mi přijde, že ten typecast má bejt na `any`, protože dokud ten parametr tu podmínku nesplní, může to bejt skutečně cokoliv..

Reply:
as any